lib/stripe/multipart_encoder.rb

```diff
@@ -0,0 +1,131 @@
+# frozen_string_literal: true
+
+require "securerandom"
+
+module Stripe
+  # Encodes parameters into a `multipart/form-data` payload as described by RFC
+  # 2388:
+  #
+  #     https://tools.ietf.org/html/rfc2388
+  #
+  # This is most useful for transferring file-like objects.
+  #
+  # Parameters should be added with `#encode`. When ready, use `#body` to get
+  # the encoded result and `#content_type` to get the value that should be
+  # placed in the `Content-Type` header of a subsequent request (which includes
+  # a boundary value).
+  class MultipartEncoder
+    MULTIPART_FORM_DATA = "multipart/form-data".freeze
+
+    # A shortcut for encoding a single set of parameters and finalizing a
+    # result.
+    #
+    # Returns an encoded body and the value that should be set in the content
+    # type header of a subsequent request.
+    def self.encode(params)
+      encoder = MultipartEncoder.new
+      encoder.encode(params)
+      encoder.close
+      [encoder.body, encoder.content_type]
+    end
+
+    # Gets the object's randomly generated boundary string.
+    attr_reader :boundary
+
+    # Initializes a new multipart encoder.
+    def initialize
+      # We seed this with an empty string so that it encoding defaults to UTF-8
+      # instead of ASCII. The empty string is UTF-8 and new string inherits the
+      # encoding of the string it's seeded with.
+      @body = String.new("")
+
+      # Chose the same number of random bytes that Go uses in its standard
+      # library implementation. Easily enough entropy to ensure that it won't
+      # be present in a file we're sending.
+      @boundary = SecureRandom.hex(30)
+
+      @closed = false
+      @first_field = true
+    end
+
+    # Gets the encoded body. `#close` must be called first.
+    def body
+      raise "object must be closed before getting body" unless @closed
+      @body
+    end
+
+    # Finalizes the object by writing the final boundary.
+    def close
+      raise "object already closed" if @closed
+
+      @body << "\r\n"
+      @body << "--#{@boundary}--"
+
+      @closed = true
+
+      nil
+    end
+
+    # Gets the value including boundary that should be put into a multipart
+    # request's `Content-Type`.
+    def content_type
+      "#{MULTIPART_FORM_DATA}; boundary=#{@boundary}"
+    end
+
+    # Encodes a set of parameters to the body.
+    #
+    # Note that parameters are expected to be a hash, but a "flat" hash such
+    # that complex substructures like hashes and arrays have already been
+    # appropriately Stripe-encoded. Pass a complex structure through
+    # `Util.flatten_params` first before handing it off to this method.
+    def encode(params)
+      raise "no more parameters can be written to closed object" if @closed
+
+      params.each do |name, val|
+        if val.is_a?(::File)
+          val.rewind
```

Comment:
Oh man, I was confused here because I added this `rewind` explicitly so that this could work with with a tempfile that I was using in the encoder's test suite.

You're totally right in that a `Tempfile` is not a `File`:

``` ruby
irb(main):001:0> require 'tempfile'
=> true
irb(main):002:0> t = Tempfile.new
=> #<Tempfile:/var/folders/b7/50czt19n47vf99rd769dq8wc0000gn/T/20190725-551-1kjvpws>
irb(main):003:0> t.is_a?(File)
=> false
```

But crazy enough, when you use `Tempfile` in its block form, the pointer you're handed _is_ a file:

``` ruby
irb(main):009:0> Tempfile.create do |f|
irb(main):010:1* f.is_a?(File)
irb(main):011:1> end
=> true
```

Which is why this code works with its test suite.

So given that, do you have thoughts on:

1.  Whether we should still whitelist `Tempfile` as well as `File` to be rewound like you suggested? (I could see it either way TBH.)
2. Whether we should `rewind` at all from the encoder? I could just change the test suite to `rewind` before handing off the tempfile like in the file tests. I was thinking that just doing a `rewind` should never be harmful, but it's possible there's some case that I hadn't though of (like if a user was purposely trying to upload only part of a file).

Reply:
> But crazy enough, when you use Tempfile in its block form, the pointer you're handed is a file:

Oh Ruby :shakes_fist:

Yeah, I'm starting to think it might be better to not call `rewind` at all and let the user do it themselves if they want. Will leave final decision up to you.

Reply:
> Oh Ruby :shakes_fist:

Haha, I know right??

Okay, thanks OB. I thought about it a little more, and in the end have decide to remove the `rewind` from the encoder and just have it present in the test suite like we were doing in the file resource tests. Doing the automatic rewind might be convenient in some cases, but it's just a little too "magic", and might obstruct the legitimate (albeit probably very rare) use case of passing in a partially read file.

I also added `Tempfile` to the same case that `File` was in where we sent a specific filename instead of just the generic word `blob`. `Tempfile` filenames will often be meaningless, but they can be given a prefix, and in cases where they have, their filename will be better than `blob`. So the code now looks like this:

``` ruby
        if val.is_a?(::File) || val.is_a?(::Tempfile)
          write_field(name, val.read, filename: ::File.basename(val.path))
        elsif val.respond_to?(:read)
          write_field(name, val.read, filename: "blob")
        else
          write_field(name, val, filename: nil)
        end
```